src/frontend/Typechecker.ml

```diff
@@ -1001,40 +1002,178 @@ let warn_self_assignment loc lhs rhs =
-      if Ast.compare_untyped_lval lhs rhs = 0 then
+      let lhs = lhs |> Ast.untyped_lvalue_of_typed_lvalue_pack in
+      if Ast.compare_untyped_lval_pack lhs rhs = 0 then
         add_warning loc "Assignment of variable to itself." )
 
 let check_assignment_operator loc assop lhs rhs =
-  let err op =
-    Semantic_error.illtyped_assignment loc op lhs.lmeta.type_ rhs.emeta.type_
+  let rec type_of_lvalue = function
+    | LValue {lmeta; _} -> lmeta.type_
+    | LTuplePack (lvs, _) -> UnsizedType.UTuple (List.map ~f:type_of_lvalue lvs)
+  in
+  let err lhs op rhs =
+    let loc =
+      match lhs with
+      | LValue ({lmeta= {loc; _}; _} : typed_lval) | LTuplePack (_, loc) -> loc
+    in
+    Semantic_error.illtyped_assignment loc op (type_of_lvalue lhs) rhs |> error
   in
   match assop with
-  | Assign | ArrowAssign -> (
+  | Assign | ArrowAssign ->
       warn_self_assignment loc lhs rhs ;
-      match
-        SignatureMismatch.check_of_same_type_mod_conv lhs.lmeta.type_
-          rhs.emeta.type_
-      with
-      | Ok p ->
-          let rhs =
-            (* Hack: need RHS to properly get promoted to var if needed *)
-            {rhs with emeta= {rhs.emeta with ad_level= lhs.lmeta.ad_level}}
-          in
-          Promotion.promote rhs p
-      | Error _ -> err Operator.Equals |> error )
+      let rec typechk lhs rhs =
+        match (lhs, rhs) with
+        | LValue {lmeta= {type_; ad_level; _}; _}, rhs -> (
+          match SignatureMismatch.check_of_same_type_mod_conv type_ rhs with
+          | Ok p -> (p, ad_level)
+          | Error _ -> err lhs Equals rhs )
+        | LTuplePack (lvs, _), UnsizedType.UTuple tps ->
+            let proms, ad_levels =
+              match List.map2 ~f:typechk lvs tps with
+              | Unequal_lengths -> err lhs Equals rhs
+              | Ok l -> List.unzip l in
+            if
+              List.exists ~f:(function NoPromotion -> false | _ -> true) proms
+            then (TuplePromotion proms, TupleAD ad_levels)
+            else (NoPromotion, TupleAD ad_levels)
+        | LTuplePack _, rhs -> err lhs Equals rhs in
+      let prom, ad_level = typechk lhs rhs.emeta.type_ in
+      let rhs =
+        (* Hack: need RHS to properly get promoted to var if needed *)
+        {rhs with emeta= {rhs.emeta with ad_level}} in
+      Promotion.promote rhs prom
   | OperatorAssign op -> (
-      let args = List.map ~f:arg_type [Ast.expr_of_lvalue lhs; rhs] in
+      let args =
+        [(UnsizedType.AutoDiffable, type_of_lvalue lhs); arg_type rhs] in
       let return_type = assignmentoperator_stan_math_return_type op args in
-      match return_type with Some Void -> rhs | _ -> err op |> error )
+      match return_type with
+      | Some Void -> rhs
+      | _ -> err lhs op rhs.emeta.type_ )
+
+let overlapping_lvalues lvals =
+  (* Prevent assigning to multiple indices in the same object at once.
+     In principal it is safe to assign to disjoint indices, but statically
```

Comment:
```suggestion
     In principle it is safe to assign to disjoint indices, but statically
```

Reply:
I swear I fixed this already. I must have copy/pasted an older version

Reply:
The funny thing is _I_ saw it when doing the `lvalue_pack` refactor and then forgot to fix it.